src/grt/src/fastroute/src/graph2d.cpp

```diff
@@ -317,4 +499,184 @@ void Graph2D::foreachEdge(const std::function<void(Edge&)>& func)
   inner(v_edges_);
 }
 
+// Initializes the 3D capacity of the graph.
+void Graph2D::initCap3D()
+{
+  v_cap_3D_.resize(boost::extents[num_layers_][x_grid_][y_grid_]);
+  h_cap_3D_.resize(boost::extents[num_layers_][x_grid_][y_grid_]);
+  initNDRnets();
+}
+
+// Initializes the NDR nets for each grid cell.
+void Graph2D::initNDRnets()
+{
+  v_ndr_nets_.resize(boost::extents[x_grid_][y_grid_]);
+  h_ndr_nets_.resize(boost::extents[x_grid_][y_grid_]);
+}
+
+// Updates the 3D capacity of a specific edge.
+void Graph2D::updateCap3D(int x,
+                          int y,
+                          int layer,
+                          EdgeDirection direction,
+                          const double cap)
+{
+  auto& cap3D = (EdgeDirection::Horizontal == direction)
+                    ? h_cap_3D_[layer][x][y]
+                    : v_cap_3D_[layer][x][y];
+  cap3D.cap = cap;
+  cap3D.cap_ndr = cap;
+}
+
+// Prints the capacity of each edge per layer.
+void Graph2D::printEdgeCapPerLayer()
+{
+  logger_->report("=== printEdgeCapPerLayer ===");
+  for (int y = 0; y < y_grid_; y++) {
+    for (int x = 0; x < x_grid_; x++) {
+      for (int l = 0; l < num_layers_; l++) {
+        if (x < x_grid_ - 1) {
+          logger_->report(
+              "\tH x{} y{} l{}: {}", x, y, l, h_cap_3D_[l][x][y].cap);
+        }
+        if (y < y_grid_ - 1) {
+          logger_->report(
+              "\tV x{} y{} l{}: {}", x, y, l, v_cap_3D_[l][x][y].cap);
+        }
+      }
+    }
+  }
+}
+
+// Checks if there is enough NDR capacity for a given net.
+bool Graph2D::hasNDRCapacity(FrNet* net, int x, int y, EdgeDirection direction)
+{
+  const int8_t edgeCost = net->getEdgeCost();
+  double max_single_layer_cap = 0;
+
+  if (edgeCost > 1) {
+    // For nets with high cost, we need at least one layer with sufficient
+    // capacity
+    for (int l = net->getMinLayer(); l <= net->getMaxLayer(); l++) {
+      double layer_cap = 0;
+      if (direction == EdgeDirection::Horizontal) {
+        layer_cap = h_cap_3D_[l][x][y].cap_ndr;
+      } else {
+        layer_cap = v_cap_3D_[l][x][y].cap_ndr;
+      }
+      max_single_layer_cap = std::max(max_single_layer_cap, layer_cap);
+      if (layer_cap >= edgeCost) {
+        return true;
+      }
+    }
+
+    // No single layer can accommodate this NDR net
+    return false;
+  }
+
+  return true;
```

Comment:
You can move the `return false` from line 574 out of the `if` block, so you don't have the awkward `return true` in the end of the function.